File: results/0.0/TIFF/tiff-deepseek-chat.c
#include <hammer/hammer.h>
#include <hammer/glue.h>
#include <stdio.h>
#include <stdlib.h>

// Define the TIFF header structure
typedef struct {
    uint16_t byte_order;
    uint16_t version;
    uint32_t ifd_offset;
} tiff_header_t;

// Define the IFD entry structure
typedef struct {
    uint16_t tag;
    uint16_t type;
    uint32_t count;
    uint32_t value_offset;
} ifd_entry_t;

// Define the IFD structure
typedef struct {
    uint16_t entry_count;
    ifd_entry_t* entries;
    uint32_t next_ifd_offset;
} ifd_t;

// Define the TIFF structure
typedef struct {
    tiff_header_t header;
    ifd_t ifd;
} tiff_t;

// Hammer parser for TIFF header
HParser* tiff_header_parser() {
    return h_sequence(
        h_int16(),  // byte_order
        h_int16(),  // version
        h_int32(),  // ifd_offset
        NULL
    );
}

// Hammer parser for IFD entry
HParser* ifd_entry_parser() {
    return h_sequence(
        h_int16(),  // tag
        h_int16(),  // type
        h_int32(),  // count
        h_int32(),  // value_offset
        NULL
    );
}

// Hammer parser for IFD
HParser* ifd_parser() {
    return h_sequence(
        h_int16(),  // entry_count
        h_many(ifd_entry_parser()),  // entries
        h_int32(),  // next_ifd_offset
        NULL
    );
}

// Hammer parser for TIFF
HParser* tiff_parser() {
    return h_sequence(
        tiff_header_parser(),
        ifd_parser(),
        NULL
    );
}

// Main function to parse TIFF
int main(int argc, char** argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <tiff_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t* buffer = (uint8_t*)malloc(file_size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParser* parser = tiff_parser();
    HParseResult* result = h_parse(parser, buffer, file_size);

    if (result) {
        printf("Parsing successful!\n");
        h_parse_result_free(result);
    } else {
        printf("Parsing failed!\n");
    }

    free(buffer);
    return 0;
}

File: results/0.0/TIFF/tiff-deepseek-chat.rs
use nom::{
    bytes::complete::{tag, take},
    number::complete::{le_u16, le_u32, le_u64},
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct TiffHeader {
    byte_order: u16,
    version: u16,
    ifd_offset: u32,
}

#[derive(Debug)]
struct IfdEntry {
    tag: u16,
    field_type: u16,
    count: u32,
    value_offset: u32,
}

#[derive(Debug)]
struct Ifd {
    entries: Vec<IfdEntry>,
    next_ifd_offset: u32,
}

#[derive(Debug)]
struct TiffFile {
    header: TiffHeader,
    ifds: Vec<Ifd>,
}

fn parse_tiff_header(input: &[u8]) -> IResult<&[u8], TiffHeader> {
    let (input, (byte_order, version, ifd_offset)) = tuple((le_u16, le_u16, le_u32))(input)?;
    Ok((
        input,
        TiffHeader {
            byte_order,
            version,
            ifd_offset,
        },
    ))
}

fn parse_ifd_entry(input: &[u8]) -> IResult<&[u8], IfdEntry> {
    let (input, (tag, field_type, count, value_offset)) =
        tuple((le_u16, le_u16, le_u32, le_u32))(input)?;
    Ok((
        input,
        IfdEntry {
            tag,
            field_type,
            count,
            value_offset,
        },
    ))
}

fn parse_ifd(input: &[u8]) -> IResult<&[u8], Ifd> {
    let (input, num_entries) = le_u16(input)?;
    let (input, entries) = nom::multi::count(parse_ifd_entry, num_entries as usize)(input)?;
    let (input, next_ifd_offset) = le_u32(input)?;
    Ok((
        input,
        Ifd {
            entries,
            next_ifd_offset,
        },
    ))
}

fn parse_tiff_file(input: &[u8]) -> IResult<&[u8], TiffFile> {
    let (input, header) = parse_tiff_header(input)?;
    let mut ifds = Vec::new();
    let mut next_ifd_offset = header.ifd_offset;

    while next_ifd_offset != 0 {
        let (remaining_input, ifd) = parse_ifd(&input[next_ifd_offset as usize..])?;
        next_ifd_offset = ifd.next_ifd_offset;
        ifds.push(ifd);
    }

    Ok((input, TiffFile { header, ifds }))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <tiff_file>", args[0]);
        return;
    }

    let mut file = File::open(&args[1]).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_tiff_file(&buffer) {
        Ok((_, tiff_file)) => {
            println!("{:#?}", tiff_file);
        }
        Err(e) => {
            eprintln!("Failed to parse TIFF file: {:?}", e);
        }
    }
}